pwiz_tools/Skyline/Test/CmdLineAssociateProteinsTest.cs

```diff
@@ -0,0 +1,93 @@
+﻿/*
+ * Original author: Nicholas Shulman <nicksh .at. u.washington.edu>,
+ *                  MacCoss Lab, Department of Genome Sciences, UW
+ *
+ * Copyright 2024 University of Washington - Seattle, WA
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Xml.Serialization;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using pwiz.Common.SystemUtil;
+using pwiz.Skyline;
+using pwiz.Skyline.Model;
+using pwiz.SkylineTestUtil;
+
+namespace pwiz.SkylineTest
+{
+    /// <summary>
+    /// Tests various combinations of "--associate-proteins-fasta" and "--save-settings" using SkylineCmd.exe
+    /// </summary>
+    [TestClass]
+    public class CmdLineAssociateProteinsTest : AbstractUnitTest
+    {
+        [TestMethod]
+        public void TestCmdLineAssociateProteins()
+        {
+            TestFilesDir = new TestFilesDir(TestContext, @"Test\CmdLineAssociateProteinsTest.zip");
+            var skylineExeFolder = Path.GetDirectoryName(typeof(SkylineWindow).Assembly.Location);
+            Assert.IsNotNull(skylineExeFolder);
+            var skylineCmdExePath = Path.Combine(skylineExeFolder, "SkylineCmd.exe");
+            AssertFileExists(skylineCmdExePath);
+
+            // Run the batch file "AssociateProteinsScript.bat". Its first argument is the
+            // path to SkylineCmd.exe
+            var batchFileName = TestFilesDir.GetTestPath("AssociateProteinsScript.bat");
+            AssertFileExists(batchFileName);
+            var workingDirectory = Path.GetDirectoryName(batchFileName);
+            Assert.IsNotNull(workingDirectory);
+            var processStartInfo = new ProcessStartInfo(batchFileName, skylineCmdExePath)
+            {
+                WorkingDirectory = workingDirectory,
+                UseShellExecute = false
+            };
+            var outputWriter = new StringWriter();
+            var processRunner = new ProcessRunner();
+            IProgressStatus status = new ProgressStatus(string.Empty);
+            processRunner.Run(processStartInfo, null, null, ref status, outputWriter);
+
```

Comment:
This is quite different from the existing CommandLine tests (which all (or almost all?) start with "Console" in the name). Why call SkylineCmd with ProcessRunner? This approach also means the batch file you're using isn't easy to track changes for. Is there some caveat about saving settings I'm not considering?

Reply:
If we keep this form, then perhaps prefixing the test with "TestSkylineCmd..." would make it clear this test actually runs SkylineCmd.

Reply:
I figured we should have a test which runs SkylineCmd.exe using a .bat file.
I was also thinking that if we wanted to test whether settings were being saved, the best way to test that would be to run SkylineCmd.exe multiple times out of process so that it actually has to read the settings from disk.
The usual way of testing the command line involves calling the command line functions in the same process as the unit tests where it would be trickier to figure out whether the "user.config" settings file was getting saved.

Reply:
I agree about running SkylineCmd.exe, I just don't think we need to hide the commands inside a batch file inside a zip file. The commands are simple enough we can keep them in the code.

Reply:
So Nick, are you proposing you'll keep the batch file in a future commit to this PR, it'll just be using the .data directory (and thus plaintext in git)? If so, I still wonder why that approach is better than just putting the commands in the code, but there's no question it's better than putting it in a zip.

Reply:
I wanted the experience of writing a .bat file which called SkylineCmd.exe.
It certainly would be possible to accomplish the same thing with multiple calls to ProcessRunner.Run, but I was thinking that we should have at least one test which uses a .bat, because that is how some users write their scripts.
Is there a reason that you think we should avoid calling SkylineCmd.exe from .bat files during our suite?

Reply:
As long as the batch script is in plain text I think it's fine, it's just a little more cumbersome to maintain than having it inline in the code. But as you say, it's useful to have at least one test which runs Skyline this way. :shipit: